File: backend/src/main/java/com/whiteblue/backend/security/util/CookieUtil.java
package com.whiteblue.backend.security.util;

import org.springframework.util.SerializationUtils;

import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.util.Arrays;
import java.util.Base64;
import java.util.Optional;

public class CookieUtil {
    public static Optional<Cookie> getCookie(HttpServletRequest request, String key) {
        Optional<Cookie[]> optionalCookies = Optional.ofNullable(request.getCookies());
        return optionalCookies.flatMap(cookies -> Arrays.stream(cookies)
                                                        .filter(cookie -> cookie.getName()
                                                                                .equals(key))
                                                        .findAny());
    }

    public static void addCookie(HttpServletResponse response, String key, String value) {
        Cookie cookie = new Cookie(key, value);
        cookie.setPath("/");
        cookie.setHttpOnly(true);
        //        cookie.setSecure(true);
        response.addCookie(cookie);
    }

    public static void deleteCookie(HttpServletRequest request, HttpServletResponse response, String key) {
        Arrays.stream(request.getCookies())
              .filter(cookie -> cookie.getName()
                                      .equals(key))
              .forEach(cookie -> {
                  cookie.setValue(null);
                  cookie.setMaxAge(0);
                  response.addCookie(cookie);
              });
    }

    public static String serialize(Object object) {
        return Base64.getUrlEncoder()
                     .encodeToString(SerializationUtils.serialize(object));
    }

    public static <T> T deserialize(Cookie cookie, Class<T> cls) {
        return cls.cast(SerializationUtils.deserialize(
                Base64.getUrlDecoder()
                      .decode(cookie.getValue())));
    }
}

File: backend/src/main/kotlin/com/whiteblue/backend/util/CookieUtil.kt
package com.whiteblue.backend.util

import jakarta.servlet.http.Cookie
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse

object CookieUtil {
    fun getCookie(request: HttpServletRequest, key: String): Cookie? = request.cookies?.find { cookie: Cookie -> cookie.name.equals(key) }

    fun addCookie(response: HttpServletResponse, key: String, value: String) = response.addCookie(Cookie(key, value)
                                                                                                      .apply {
                                                                                                          path = "/"
                                                                                                          isHttpOnly = true
                                                                                                          secure = true
                                                                                                      })

    fun deleteCookie(request: HttpServletRequest, response: HttpServletResponse, key: String) =
        request.cookies?.filter { it.name.equals(key) }
            ?.forEach {
                it.value = null
                it.maxAge = 0
                response.addCookie(it)
            }
}
